### ISSUE TITLE
Crash on Object to RLMObject after ObjectiveCSupport.convert

### ISSUE THREAD

**Issue by liuxuan30** [realm/realm-cocoa#5555]
## Problem
There is a framework https://github.com/danielgindi/ChartsRealm, branch `Generic`

In the framework, there is some init() like

```swift
    public convenience init<T: Object>(results: Results<T>?, xValueField: String?, yValueField: String)
    {
        var converted: RLMResults<RLMObject>?
        
        if results != nil
        {
            converted = ObjectiveCSupport.convert(object: results!) as? RLMResults<RLMObject>
        }
        
        self.init(results: converted, xValueField: xValueField, yValueField: yValueField)
    }
```
It uses `ObjectiveCSupport.convert(object: results!)` to convert `Object<T>` to `RLMResults<RLMObject>`

However when I generate some Results<SomeObject> and call init(), it eventually crashed at:
in `@objc internal func buildCache()` at: 
>buildEntryFromResultObject(e as! RLMObject, x: xValue)

```swift
    @objc internal func buildCache()
    {
        // _results is @objc internal var _results: RLMResults<RLMObject>?
        guard let results = _results else { return }  
        
        _cache.removeAll()
        _cache.reserveCapacity(Int(results.count))
        
        var xValue: Double = 0.0
        
        var iterator = NSFastEnumerationIterator(results)
        while let e = iterator.next()
        {
            _cache.append(buildEntryFromResultObject(e as! RLMObject, x: xValue))
            xValue += 1.0
        }
    }
```
saying:
>Could not cast value of type 'SomeObject' (0x6040001f6710) to 'RLMObject' (0x109aeb940).

## Expected Results
does not crash

## Actual Results
crash

## Steps to Reproduce
download https://github.com/danielgindi/ChartsRealm, branch `Generic`

create 
```swift
import RealmSwift
class SomeObject: Object
{
    @objc dynamic var datetime = ""
    @objc dynamic var city = 0
}
```

Try to generate some data into Realm, and make a query in swift, like
`let results = realm.objects(SomeObject.self).filter("city=\(i + 1)")`

pass this `results` to:
`RealmLineDataSet(results: results, xValueField: xValueField, yValueField: yValueField)`

and it will crash later.

## Version of Realm and Tooling
Realm framework version: latest master branch
Xcode version: 9.2 Swift 4.0.3
iOS/OSX version: iOS 11, 
Dependency manager + version: carthage latest

**Comment by liuxuan30** [comment]
No one can take a quick a look?

**Comment by bdash** [comment]
I'm not quite sure what you expect to happen here. If `SomeObject` is subclass of `RealmSwift.Object`, it does not inherit from `RLMObject`, so the cast should fail.

**Comment by liuxuan30** [comment]
From the definition:
```swift
    /// Convert a `Results` to a `RLMResults`.
    public static func convert<T>(object: RealmSwift.Results<T>) -> RLMResults<Swift.AnyObject> where T : RealmCollectionValue
```
`Object` is supposed to conform `RealmCollectionValue`:
```swift
@objc(RealmSwiftObject) open class Object : RLMObjectBase, ThreadConfined, RealmCollectionValue {
```
So I think it's a runtime crash bug.

**Comment by bdash** [comment]
The problem is your assumption that the objects in the resulting `RLMResults` are instances of subclasses of `RLMObject`. They’re not. Their type doesn’t change just because you’re accessing them via `RLMResults` rather than `Results`. Attempting to cast an `Object` subclass to `RLMObject` will always fail since the types are not related.

**Comment by liuxuan30** [comment]
So about the old implementation, it actually works. I will check how the realm data values are generated in the first place.

**Comment by liuxuan30** [comment]
Hi @bdash  you are right. I found that ChartsRealm only demo is written in Objc, so it just takes RLMObject as the demo data values, like you said, the conversion will pass.
It looks like the swift API interface is not tested and takes the assumption the conversion of Object will pass.

Now, may I ask do you have any suggestions? Like if I want to make ChartRealm to support both Swift and Objc, and route all interfaces finally to the same one (like routing Swift to Objc implementation), is it possible?

I think the main problem is that are there any methods can convert `Object` (or plus `RLMObject`) to eventually a Objc compatible one?

**Comment by bdash** [comment]
`Object` and `RLMObject` have a common base type: `RLMObjectBase`. I'd imagine you could take advantage of this to support objects of both types.

**Comment by liuxuan30** [comment]
Thanks! Closing this, as it seems ChartsRealm issue.

**Comment by liuxuan30** [comment]
@bdash  sorry to ask some more questions.

I found using `RLMObjectBase` seems hard. As many of the APIs require Object or RLMObject, and people usually build on these two types. So basically I have to provide implementations for both Swift and Objc.

But I suddenly start to wonder why using RLMObject in the first place.

So the question is, can RealmSwift works in Objc environment? If it could work and I add necessary `@objc` tags to expose the methods, I could abandon using RLMObject then.

Do you think RealmSwift will perform worse than Objc, or have any side effects?

**Comment by liuxuan30** [comment]
Just  found generic does not work on Objc. Seems a dead end but to reimplement

**Comment by liuxuan30** [comment]
@bdash Sorry I must be tried and get distracted. The real question I want to ask is:

why 
```swift
converted = ObjectiveCSupport.convert(object: results!) as? RLMResults<RLMObject>
```
work on `<T: Object>(results: Results<T>?)` type?

It is said that generic types are different types, but `converted` is valid after the conversion, even the element type is not RLMObject:
```
(lldb) po converted!
RLMResults<AirQuality> <0x7fe63460d2a0> (
	[0] AirQuality {
		_datetime = 2018-01-17 03:00:00 +0000;
		_city = 1;
		_compoundKey = 1-2018-01-17 03:00:00 +0000;
		concentration = 69;
		AQI = 158;
```
`AirQuality` is sub class of `Object`

**Comment by bdash** [comment]
Objective-C generics are a compile-time construct, the generic isn't part of the runtime type. This means it's possible to successfully cast a `RLMResults` to a `RLMResults` constrained on a completely unrelated type without error, since the only part of the type information that's available at runtime is the same.

**Comment by liuxuan30** [comment]
Hi @bdash , thanks for reply. 
I'm now confused. I used to ask a question 
[RealmSwift Cannot cast Results<SomeOjbect> to Results<Object>](https://stackoverflow.com/questions/48125923/realmswift-cannot-cast-resultssomeojbect-to-resultsobject/48126909?noredirect=1#comment83231956_48126909)

It's said `Results<SomeOjbect>` and `Results<Object>` are different types totally, even SomeOjbect is sub class of Object. so I cannot do cast like `as? Results<Object>`, I must use constraints.

However, here I could eventually do the cast: `as? RLMResults<RLMObject>`:
```swift
converted = ObjectiveCSupport.convert(object: results!) as? RLMResults<RLMObject>
```
I checked convert method:
```swift
    /// Convert a `Results` to a `RLMResults`.
    public static func convert<T>(object: RealmSwift.Results<T>) -> RLMResults<Swift.AnyObject> where T : RealmCollectionValue
```
Why we could cast `RLMResults<Swift.AnyObject>` to `RLMResults<RLMObject>` here?

What did I miss? Or Realm conformed any protocol can do this?
